dc/s2n-quic-dc/src/stream/client/tokio.rs

```diff
@@ -60,7 +64,14 @@ where
     Sub: event::Subscriber,
 {
     // Race TCP handshake with the TLS handshake
-    let (socket, peer) = tokio::try_join!(TcpStream::connect(acceptor_addr), handshake,)?;
+    let handshake = async {
+        let peer = handshake.await?;
+        let (crypto, parameters) = peer.pair(&TransportFeatures::TCP);
+        Ok((peer, crypto, parameters))
+    };
+    // poll the crypto first so the server can read the first packet on accept in the happy path
```

Comment:
One thought I just had: this does mean that if the acceptor address is variable (e.g., because there are more than one server sharing the same endpoint) and it *doesn't* respond, we'd introduce a gap in our key ID sequence where previously that wouldn't happen.

I think that gap is acceptable though. If we wanted to avoid it we could probably try to have a tiny cache of derived keys, but that seems expensive in memory and not worth the hassle (at least until we actually have problems here).

Reply:
The crypto was in open_stream, right? So I think this is a marked change (including failed opens and slow opens). If it was as you say then our original attempt of just flipping order would have sufficed in practice (since we always hit the would block with tokio today).

Well, more accurately, tokio doesn't check that the stream is open immediately - that is deferred to epoll. Connect in non-blocking mode doesn't actually get called more than once.

Reply:
> The crypto was in open_stream, right?

Ah you're correct. Yeah that's a good callout.